sql/schema/dsl.go

```diff
@@ -0,0 +1,583 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package schema
+
+import (
+	"reflect"
+)
+
+// The functions and methods below provide a DSL for creating schema resources using
+// a fluent interface. Note, methods named "Add<T>" links the arguments in both directions.
+
+// New creates a new Schema.
+func New(name string) *Schema {
+	return &Schema{Name: name}
+}
+
+// SetCharset sets or appends the Charset attribute
+// to the schema with the given value.
+func (s *Schema) SetCharset(v string) *Schema {
+	replaceOrAppend(&s.Attrs, &Charset{V: v})
+	return s
+}
+
+// SetCollation sets or appends the Collation attribute
+// to the schema with the given value.
+func (s *Schema) SetCollation(v string) *Schema {
+	replaceOrAppend(&s.Attrs, &Collation{V: v})
+	return s
+}
+
+// SetComment sets or appends the Comment attribute
+// to the schema with the given value.
+func (s *Schema) SetComment(v string) *Schema {
+	replaceOrAppend(&s.Attrs, &Comment{Text: v})
+	return s
+}
+
+// AddAttrs adds and additional attributes to the schema.
+func (s *Schema) AddAttrs(attrs ...Attr) *Schema {
+	s.Attrs = append(s.Attrs, attrs...)
+	return s
+}
+
+// SetRealm sets the database/realm of the schema.
+func (s *Schema) SetRealm(r *Realm) *Schema {
+	s.Realm = r
+	return s
+}
+
+// AddTables adds and links the given tables to the schema.
+func (s *Schema) AddTables(tables ...*Table) *Schema {
+	for _, t := range tables {
+		t.SetSchema(s)
+	}
+	s.Tables = append(s.Tables, tables...)
+	return s
+}
+
+// NewTable creates a new Table.
+func NewTable(name string) *Table {
+	return &Table{Name: name}
+}
+
+// SetCharset sets or appends the Charset attribute
+// to the table with the given value.
+func (t *Table) SetCharset(v string) *Table {
+	replaceOrAppend(&t.Attrs, &Charset{V: v})
+	return t
+}
+
+// SetCollation sets or appends the Collation attribute
+// to the table with the given value.
+func (t *Table) SetCollation(v string) *Table {
+	replaceOrAppend(&t.Attrs, &Collation{V: v})
+	return t
+}
+
+// SetComment sets or appends the Comment attribute
+// to the table with the given value.
+func (t *Table) SetComment(v string) *Table {
+	replaceOrAppend(&t.Attrs, &Comment{Text: v})
+	return t
+}
+
+// AddChecks appends the given checks to the attribute list.
+func (t *Table) AddChecks(checks ...*Check) *Table {
+	for _, c := range checks {
+		t.Attrs = append(t.Attrs, c)
+	}
+	return t
+}
+
+// SetSchema sets the schema (named-database) of the table.
+func (t *Table) SetSchema(s *Schema) *Table {
+	t.Schema = s
+	return t
+}
+
+// SetPrimaryKey sets the primary-key of the table.
+func (t *Table) SetPrimaryKey(pk *Index) *Table {
```

Comment:
In this case you do a "link-back" even though the method is not named `AddT`

Reply:
Right. The comment at the top was written before I wrote this method.